x/sequencer/keeper/sequencer.go

```diff
@@ -172,3 +146,127 @@ func (k Keeper) removeUnbondingSequencer(ctx sdk.Context, sequencer types.Sequen
 	unbondingQueueKey := types.UnbondingSequencerKey(sequencer.Address, sequencer.UnbondTime)
 	store.Delete(unbondingQueueKey)
 }
+
+/* -------------------------------------------------------------------------- */
+/*                                notice period                               */
+/* -------------------------------------------------------------------------- */
+
+// GetMatureNoticePeriodSequencers returns all sequencers that have finished their notice period
+func (k Keeper) GetMatureNoticePeriodSequencers(ctx sdk.Context, endTime time.Time) (list []types.Sequencer) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := store.Iterator(types.NoticePeriodQueueKey, sdk.PrefixEndBytes(types.NoticePeriodQueueByTimeKey(endTime)))
+
+	defer iterator.Close() // nolint: errcheck
+
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.Sequencer
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		list = append(list, val)
+	}
+
+	return
+}
+
+// AddSequencerToNoticePeriodQueue set sequencer in notice period queue
+func (k Keeper) AddSequencerToNoticePeriodQueue(ctx sdk.Context, sequencer types.Sequencer) {
+	store := ctx.KVStore(k.storeKey)
+	b := k.cdc.MustMarshal(&sequencer)
+
+	noticePeriodKey := types.NoticePeriodSequencerKey(sequencer.Address, sequencer.NoticePeriodTime)
+	store.Set(noticePeriodKey, b)
+}
+
+// remove sequencer from notice period queue
+func (k Keeper) removeNoticePeriodSequencer(ctx sdk.Context, sequencer types.Sequencer) {
+	store := ctx.KVStore(k.storeKey)
+	noticePeriodKey := types.NoticePeriodSequencerKey(sequencer.Address, sequencer.NoticePeriodTime)
+	store.Delete(noticePeriodKey)
+}
+
+/* ------------------------- proposer/next proposer ------------------------- */
+
+// GetAllProposers returns all proposers for all rollapps
+func (k Keeper) GetAllProposers(ctx sdk.Context) (list []types.Sequencer) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ProposerByRollappKey(""))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close() // nolint: errcheck
+
+	for ; iterator.Valid(); iterator.Next() {
+		address := string(iterator.Value())
+		seq := k.MustGetSequencer(ctx, address)
+		list = append(list, seq)
+	}
+
+	return
+}
+
+func (k Keeper) SetProposer(ctx sdk.Context, rollappId, sequencerAddr string) {
+	store := ctx.KVStore(k.storeKey)
+	addressBytes := []byte(sequencerAddr)
+
+	activeKey := types.ProposerByRollappKey(rollappId)
+	store.Set(activeKey, addressBytes)
+}
+
+// GetProposer returns the proposer for a rollapp
+func (k Keeper) GetProposer(ctx sdk.Context, rollappId string) (val types.Sequencer, found bool) {
+	store := ctx.KVStore(k.storeKey)
+	b := store.Get(types.ProposerByRollappKey(rollappId))
+	if len(b) == 0 {
+		return val, false
+	}
+
+	return k.GetSequencer(ctx, string(b))
+}
+
+func (k Keeper) removeProposer(ctx sdk.Context, rollappId string) {
+	k.SetProposer(ctx, rollappId, NO_SEQUENCER_AVAILABLE)
+}
```

Comment:
maybe use same pattern for next propsoer?

Reply:
discussed already. for nextProposer there's a difference between set/unset and set with ""